libqaul/http-api/src/auth/login.rs

```diff
@@ -0,0 +1,104 @@
+use crate::{
+    JsonApi,
+    models::{
+        UserAuth,
+        UserGrant,
+    },
+    QaulCore,
+    JSONAPI_MIME,
+};
+use chrono::{ DateTime, offset::Utc };
+use libqaul::UserAuth as QaulUserAuth;
+use iron::{
+    prelude::*,
+    status::Status,
+};
+use json_api::{
+    Document,
+    OptionalVec,
+    ResourceObject,
+};
+use std::convert::TryInto;
+use super::{
+    AuthError,
+    Authenticator,
+    CurrentUser,
+};
+
+pub fn login(req: &mut Request) -> IronResult<Response> {
+    // data should contain exactly one object
+    let data = match &req.extensions.get::<JsonApi>().unwrap().data {
+        OptionalVec::One(Some(d)) => d,
+        OptionalVec::Many(_) => { 
+            return Err(AuthError::MultipleData.into());
+        },
+        _ => {
+            return Err(AuthError::NoData.into());
+        },
+    };
+
+    // try to decode the payload
+    let ua : ResourceObject<UserAuth> = match data.try_into() {
+        Ok(ua) => ua,
+        Err(e) => {
+            return Err(AuthError::ConversionError(e).into());
+        },
+    };
+
+    // is the identity valid
+    let (identity, secret) = match UserAuth::into_identity(ua) {
+        Ok(id) => id,
+        Err(e) => {
+            return Err(AuthError::InvalidIdentity(e).into());
+        },
+    };
+
+    // is there a secret (there has to be a secret!)
+    let secret = match secret {
+        Some(s) => s,
+        None => { 
+            return Err(AuthError::NoSecret.into());
+        },
+    };
+
+    let qaul = req.extensions.get::<QaulCore>().unwrap();
+
+    // perform the login
+    let (ident, token) = match qaul.user_login(identity.clone(), &secret) {
+        Ok(QaulUserAuth::Trusted(ident, token)) => (ident, token),
+        Ok(QaulUserAuth::Default(_)) => {
```

Comment:
Maybe just put the `unreachable!()` on the next line. Maybe we should also rename the `UserAuth` variant here to be more self descriptive that makes the comment obsolete

Reply:
i renamed it to `UserAuth::Untrusted` which hopefully better reflects why that is a meaningless return